# Exploratory notebook

In [1]:
# Import Libraries
import pandas as pd
import numpy as np
import os
import urllib.parse
import gzip
import seaborn as sns
import matplotlib.pyplot as plt
import scipy.stats as stats
from scipy.stats import chi2_contingency
import explore_ms as e
import env
from sklearn.ensemble import IsolationForest

from io import BytesIO
from io import StringIO
from sqlalchemy import create_engine

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, precision_score, recall_score
from sklearn.tree import DecisionTreeClassifier, plot_tree
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report
from IPython.display import display, display_html
from sklearn.dummy import DummyClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import LabelEncoder
from sklearn.ensemble import RandomForestClassifier

# ACQUIRE
- Data acquired from Codeup data base 
- SQL query joined neccesary data only for data set 
- DATA TYPE SUMMARY
- 2 integer data types 
- 6 object data type 
- 0 null values


In [2]:
logs_df = e.get_sql_data()

In [3]:
logs_df.head().T

0             1                                    2  \
lesson                 /       java-ii  java-ii/object-oriented-programming   
user_id                1             1                                    1   
cohort           Hampton       Hampton                              Hampton   
program_id             1             1                                    1   
ip          97.105.19.61  97.105.19.61                         97.105.19.61   
start_date    2015-09-22    2015-09-22                           2015-09-22   
end_date      2016-02-06    2016-02-06                           2016-02-06   
program          web dev       web dev                              web dev   

                                             3                          4  
lesson      slides/object_oriented_programming  javascript-i/conditionals  
user_id                                      1                          2  
cohort                                 Hampton                      Teddy  
program_id                                   1                          2  
ip                                97.105.19.61               97.105.19.61  
start_date                          2015-09-22                 2018-01-08  
end_date                            2016-02-06                 2018-05-17  
program                                web dev                    web dev

In [4]:
logs_df.tail().T

847325             847326                   847327  \
lesson      jquery/personal-site  jquery/mapbox-api  jquery/ajax/weather-map   
user_id                       64                 64                       64   
cohort                     Staff              Staff                    Staff   
program_id                     2                  2                        2   
ip                 71.150.217.33      71.150.217.33            71.150.217.33   
start_date            2014-02-04         2014-02-04               2014-02-04   
end_date              2014-02-04         2014-02-04               2014-02-04   
program                  web dev            web dev                  web dev   

                                                      847328  \
lesson      anomaly-detection/discrete-probabilistic-methods   
user_id                                                  744   
cohort                                                 Staff   
program_id                                                 2   
ip                                             24.160.137.86   
start_date                                        2014-02-04   
end_date                                          2014-02-04   
program                                              web dev   

                       847329  
lesson      jquery/mapbox-api  
user_id                    64  
cohort                  Staff  
program_id                  2  
ip              71.150.217.33  
start_date         2014-02-04  
end_date           2014-02-04  
program               web dev

In [5]:
logs_df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 847330 entries, 0 to 847329
Data columns (total 8 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   lesson      847329 non-null  object
 1   user_id     847330 non-null  int64 
 2   cohort      847330 non-null  object
 3   program_id  847330 non-null  int64 
 4   ip          847330 non-null  object
 5   start_date  847330 non-null  object
 6   end_date    847330 non-null  object
 7   program     847330 non-null  object
dtypes: int64(2), object(6)
memory usage: 51.7+ MB


In [6]:
logs_df.columns.tolist()


['lesson',
 'user_id',
 'cohort',
 'program_id',
 'ip',
 'start_date',
 'end_date',
 'program']

In [7]:
logs_df.shape

(847330, 8)

In [8]:
logs_df.describe()


user_id     program_id
count  847330.000000  847330.000000
mean      456.707344       2.086004
std       250.734201       0.388231
min         1.000000       1.000000
25%       263.000000       2.000000
50%       476.000000       2.000000
75%       648.000000       2.000000
max       981.000000       4.000000

# PREPARE
- Data Cleaning:
- Drop unnecessary axis
- Rename
- Find nulls
- Drop nulls
- Check preperation
- The data set has 8 columns and 847330 rows

In [9]:
logs_df.isnull().sum()


lesson        1
user_id       0
cohort        0
program_id    0
ip            0
start_date    0
end_date      0
program       0
dtype: int64

In [10]:
logs_df['lesson'].value_counts()

/                                                               45854
javascript-i                                                    18203
toc                                                             17591
search/search_index.json                                        17534
java-iii                                                        13166
                                                                ...  
content/examples/javascript/primitive-types.html                    1
content/examples/javascript/conditionals.html                       1
2-storytelling/1-overview/www.qlik.com                              1
syntax-types-and-variables                                          1
appendix/professional-development/post-interview-review-form        1
Name: lesson, Length: 2224, dtype: int64

In [11]:
logs_df['cohort'].value_counts().head()


Staff      84031
Ceres      40730
Zion       38096
Jupiter    37109
Fortuna    36902
Name: cohort, dtype: int64

In [12]:
logs_df['program'].value_counts().head()


web dev         743913
data science    103412
frontend             5
Name: program, dtype: int64

# EXPLORE

- Explored lesson access patterns across programs.
- Detected least accessed lessons, including "Introduction to GitHub."
- Advised enhancing content with practical examples.
- Proposed integrating real-world context for better engagement.
- Recommended advanced lessons on branch pruning and merging.
- Insights guide curriculum improvements for enhanced learning outcomes.


### <span style ='color:#D16002'>1. Which lesson appears to attract the most traffic consistently across cohorts (per program)?

In [13]:
# Call the function to get top lessons for web dev and cohorts
top_lessons_program_wd, top_lesson_across_cohorts_wd = e.get_top_lessons_by_program(logs_df, program='web dev')
"Top Lessons for Program web development:"
top_lessons_program_wd

print("Top Lesson Across Cohorts for Program Web Development:")
print(top_lesson_across_cohorts_wd)
 

Top Lesson Across Cohorts for Program Web Development:
html-css


In [14]:
# Call the function to get top lessons for data science across cohorts
top_lessons_program_ds, top_lesson_across_cohorts_ds = e.get_top_lessons_by_program(logs_df, program='data science')
"Top Lessons for Program data science:"
top_lessons_program_ds

print("Top Lesson Across Cohorts for Program Data  Science:")
print(top_lesson_across_cohorts_ds)


Top Lesson Across Cohorts for Program Data  Science:
1-fundamentals/1.1-intro-to-data-science


In [15]:
# Call the function to get top lessons for data science across cohorts
top_lessons_program_fe, top_lesson_across_cohorts_fe = e.get_top_lessons_by_program(logs_df, program='frontend')
"Top Lessons for Program Frontend:"
top_lessons_program_fe

print("Top Lesson Across Cohorts for Program Front-end")
print(top_lesson_across_cohorts_fe)


Top Lesson Across Cohorts for Program Front-end
content/html-css



###  <span style ='color:#D16002'> QUESTION 1 Findings

#### <span style ='color:#D16002'> **Methodology and Insights:**<span style= 'color: #EC9706'> Outliers were removed to prevent data skewness and ensure more accurate insights.Data Grouping Grouped by program, lesson, and cohort to gauge lesson popularity within cohorts.Data Sorting Lessons were sorted based on student access frequency.Program Filtering focus was narrowed down to specific program-related lessons.Irrelevant Lesson Removal lessons not pertinent to the program or not favored by students were excluded.

####  <span style ='color:#D16002'>**Findings: Top Lesson Across Cohorts for Different Programs**

- <span style= 'color: #EC9706'> **Web Development Program:** Are all related to the fundamentals of web development, such as HTML, CSS, and JavaScript. This is because these are the skills that all web developers need to know.

- <span style= 'color: #EC9706'> **Data Science Program:** Are more focused on advanced topics, such as statistics, machine learning, and natural language processing. This is because these are the skills that data scientists need to know in order to analyze data and solve real-world problems.

- <span style= 'color: #EC9706'> **Frontend program:** Are a mix of fundamentals and advanced topics. This is because frontend developers need to know both the fundamentals of web development, such as HTML and CSS, as well as more advanced topics, such as React and Node.js.

#### <span style ='color:#D16002'> **Takeaways:**
#### <span style= 'color: #EC9706'>Overall, the top lessons across cohorts provide a good overview of the skills that students need to learn in order to be successful in each program. This information can be used to improve the curriculum and make it more relevant and engaging to students.

#### <span style ='color:#D16002'> **Recommendation:**
####  <span style= 'color: #EC9706'>Leverage the robust methodology and insights to fine-tune curriculum alignment. Customize lessons to program-specific needs, encompassing both fundamentals and advanced topics. Enhance engagement with real-world applications, projects, and interactive discussions. Additionally, consider investigating whether top lesson engagement is influenced by the level of difficulty or test-oriented engagement, as tests are open notes. This comprehensive approach will empower dynamic learning experiences, addressing skill relevance and potential engagement nuances.
 

### A function was created to identify the most trafficked lesson per Program accross Cohorts and these are the results 

### <span style ='color:#D16002'>7. Which lessons are least accessed?


In [30]:
# Call the function to get top lessons for web dev and cohorts
least_lessons_program_wd, least_lesson_across_cohorts_wd = e.get_least_accessed_lessons(logs_df, program='web dev')
"Least Accessed Lessons for Program web development:"
least_lessons_program_wd

print("Least Accessed Lessons Across Cohorts for Program Web Development:")
print(least_lesson_across_cohorts_wd)

Least Accessed Lessons Across Cohorts for Program Web Development:
lesson
'                                 1767
.git                              1768
.gitignore                        1769
.well-known/assetlinks.json       1770
00_                               1812
                                 ...  
web-design/ux/purpose            13560
web-dev-day-two                  13561
working-with-time-series-data    13562
wp-admin                         13563
wp-login                         13564
Name: count, Length: 2048, dtype: int64


In [29]:
# Get the least accessed lessons for the 'data science' program
ds_least_accessed, ds_least_overall = e.get_least_accessed_lessons(logs_df, program='data science')
print("Least Accessed Lessons for Data Science Program:")
print(ds_least_accessed)
print("\nOverall Least Accessed Lesson for Data Science across all cohorts:")
print(ds_least_overall)



Least Accessed Lessons for Data Science Program:
           program                           lesson    cohort  count
0     data science  %20https://github.com/RaulCPena     Curie      1
1728  data science      timeseries/modeling-lesson2  Florence      1
1710  data science                        tidy-data     Bayes      1
1705  data science          storytelling/understand     Curie      1
1700  data science             storytelling/tableau     Curie      1
1699  data science             storytelling/tableau     Bayes      1
1693  data science              storytelling/refine     Curie      1
1692  data science              storytelling/refine     Bayes      1
1689  data science             storytelling/present     Curie      1
1688  data science             storytelling/present     Bayes      1

Overall Least Accessed Lesson for Data Science across all cohorts:
lesson
%20https://github.com/RaulCPena                0
,%20https://github.com/RaulCPena               1
.gitignore         

In [28]:
# Get the least accessed lessons for the 'Front-end' program
ds_least_accessed, ds_least_overall = e.get_least_accessed_lessons(logs_df, program='frontend')
print("Least Accessed Lessons for Front-end Program:")
print(ds_least_accessed)
print("\nOverall Least Accessed Lesson for Front-end across all cohorts:")
print(ds_least_overall)



Least Accessed Lessons for Front-end Program:
       program                                       lesson  cohort  count
1766  frontend           content/html-css/introduction.html  Apollo      1
1765  frontend  content/html-css/gitbook/images/favicon.ico  Apollo      1
1764  frontend                             content/html-css  Apollo      2

Overall Least Accessed Lesson for Front-end across all cohorts:
lesson
content/html-css                               1764
content/html-css/gitbook/images/favicon.ico    1765
content/html-css/introduction.html             1766
Name: count, dtype: int64


###  <span style ='color:#D16002'> QUESTION 7 Findings

#### <span style ='color:#D16002'> **Methodology and Insights:**<span style= 'color: #EC9706'> Remove outliers this is done to ensure that the results are not skewed by a small number of extreme values.Group the data by program, lesson, and cohort. This allows us to see how many students have accessed each lesson in each cohort.Sort the data by the number of students who have accessed each lesson. This gives us a list of the least accessed lessons.Filter the data allows us to focus on the lessons that are relevant to the specific program.Remove any unwanted lessons. This can include lessons that are not relevant to the program or that are not well-received by students and data with special characters were removed.



####  <span style ='color:#D16002'>**Findings:** Least Accessed Lesson
#### <span style= 'color: #EC9706'>The date range of the data to see how the popularity of lessons changes over time.The level of engagement with the lessons to measure how students are interacting with the content. The lessons that attract the least attention across the cohorts per program are the ones that are least insight or relevance to engaging to students.

#### <span style ='color:#D16002'>**Takeaways:** 
<span style= 'color: #EC9706'>As you can see, These findings reveal the least accessed lessons across various programs and the overall dataset. Notably, the lesson "Introduction to GitHub" emerged as the least accessed, reflecting minimal engagement across all programs and cohorts.
    

#### <span style ='color:#D16002'>**Recommendation:**
####  <span style= 'color: #EC9706'>Considering the limited engagement with the "Introduction to GitHub" lesson across all programs and cohorts, a tailored approach is advised. Enhancing the lesson's content delivery, incorporating practical examples, and highlighting the benefits of version control could encourage greater student interaction. Moreover, to reinforce students' proficiency in GitHub usage, introducing additional lessons that delve into advanced topics, such as branch pruning and merging techniques, could be immensely valuable. By guiding students through these essential practices, the curriculum would empower them to effectively manage their repositories, collaborate seamlessly, and enhance their overall understanding of version control workflows.


 